csharp/src/Google.Protobuf.Test/CodedInputStreamAsyncTest.cs

```diff
@@ -0,0 +1,502 @@
+#region Copyright notice and license
```

Comment:
The main problem is that async code has completely different semantics (continuation) despite it looks very similar. This is the most important practical limitation of async-await: to make it efficient you need to rewrite everything from the top to the bottom using continuations. Whenever you are using an sync-async bridge like: GetAwaiter().GetResult() or using IO-bound sync operation from async code you are blocking a threadpool thread.

However if you have any ideas how to to this I would be happy to improve this code.